### PR TITLE
Initial commit of libpng-1.6.21

### PR DESCRIPTION
**Pull request by jimklimov**
Just a clone of libpng14 recipe.

### CODE REVIEW

**Comment by pyhalov** [comment]
1) You should ensure that all libpng dependencies build with libpng 1.6.
2) Mediated library is not the best thing, just preserve the latest build link.
3) Update meta-packages/libpng

**Comment by jimklimov** [comment]
Re2: Do you mean to comment away the older versions of basename "libpng.so" and not mediate the newest, like before? What if a user only installs e.g. "libpng12" package - and then has no "libpng.so" at all?

**Comment by pyhalov** [comment]
he is not supposed to do this... Perhaps, individual packages should depend on meta-package.

**Comment by jimklimov** [comment]
Interesting... now looked at the meta-package - it is wrong (refers to 1.4.17, and we have 1.4.19 component in the gate before this PR). Probably such de-coupling is not healthy when you don't know about it (e.g. a comment in child package Makefile?)

Apparently this is similar to what I did with zeromq.p5m files - although updating those regularly with the current git-head packaging (what the project suggests) and double-accounting the newest version in the actual package and in metapackage located far away would be complicated.

Is it possible to automate such upgrades (at least within the baseline version like "libpng14" for this case) by including the actual package's Makefile and overriding the build, install etc. recipes in the metapackage Makefile, so effectively we only have the IPS info and no build recipe?

**Comment by pyhalov** [comment]
Actually, meta-package doesn't need correct version :) It should just have version > than previous package with this name

**Comment by jimklimov** [comment]
Trial by fire fixed a few bugs - at least the new libgd2 (in PR) builds against installed libpng16 with co-installed libpng14 (mediated) and older libpng12+10 from upstream repo (those with files commented away) ;)

**Comment by jimklimov** [comment]
Collapsed into one commit...

File: components/library/libpng12/libpng12.p5m [hunk 1]
```diff
@@ -22,33 +22,25 @@ set name=org.opensolaris.consolidation value=$(CONSOLIDATION)
 
 license $(COMPONENT_LICENSE_FILE) license='$(COMPONENT_LICENSE)'
 
-# Commented out files should be delivered by the most recent
-# libpng version (currently it is libpng 1.4)
+# Certain files should be delivered by the most recent
```
Comment (pyhalov) [inline components/library/libpng12/libpng12.p5m]: Leave original comment and avoid removing commented out files - this is a kind of documentation

File: components/library/libpng12/libpng12.p5m [hunk 2]
```diff
@@ -22,33 +22,25 @@ set name=org.opensolaris.consolidation value=$(CONSOLIDATION)
...
 file path=usr/lib/libpng.so.3.56.0
 link path=usr/lib/libpng12.so target=libpng12.so.0.56.0
 link path=usr/lib/libpng12.so.0 target=libpng12.so.0.56.0
 file path=usr/lib/libpng12.so.0.56.0
-#link path=usr/lib/pkgconfig/libpng.pc target=libpng12.pc
 file path=usr/lib/pkgconfig/libpng12.pc
-#file path=usr/share/man/man3/libpng.3
-#file path=usr/share/man/man3/libpngpf.3
-#file path=usr/share/man/man5/png.5
+file usr/share/man/man3/libpng.3 path=usr/share/man/man3/libpng.3.libpng12
```
Comment (pyhalov) [inline components/library/libpng12/libpng12.p5m]: Don't deliver manpages for old libpng versions

File: components/library/libpng14/libpng14.p5m [hunk 3]
```diff
@@ -22,26 +22,21 @@ set name=org.opensolaris.consolidation value=$(CONSOLIDATION)
 
 license $(COMPONENT_LICENSE_FILE) license='$(COMPONENT_LICENSE)'
 
-link path=usr/bin/$(MACH64)/libpng-config target=libpng14-config
+# Certain files should be delivered by the most recent
+# libpng version (currently it is libpng 1.6)
+
 file path=usr/bin/$(MACH64)/libpng14-config
-link path=usr/bin/libpng-config target=libpng14-config
```
Comment (pyhalov) [inline components/library/libpng14/libpng14.p5m:27]: Leave comment in place and comment out common files. Also comment out manpages

File: components/library/libpng16/libpng16.p5m [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+set name=info.upstream-url value=$(COMPONENT_PROJECT_URL)
+set name=info.source-url value=$(COMPONENT_ARCHIVE_URL)
+set name=org.opensolaris.consolidation value=$(CONSOLIDATION)
+
+license $(COMPONENT_LICENSE_FILE) license='$(COMPONENT_LICENSE)'
+
+# Certain files should be delivered by the most recent
+# libpng version (currently it is libpng 1.6) as links:
+
+link path=usr/bin/$(MACH64)/libpng-config target=libpng16-config
```
Comment (pyhalov) [inline components/library/libpng16/libpng16.p5m]: Deliver manpages directly, without links

**Comment by jimklimov** [comment]
Recent PR comments were taken into account - libpng10 and libpng12 are now essentially untouched, libpng14 manifest contains the commented-away files and some more comments, and libpng16 recipe is restructured a bit; "gmake publish" still passes ;)

**Comment by pyhalov** [comment]
This looks good, you can merge commits. However, I still have to check that all libpng dependencies can be published.

**Comment by pyhalov** [comment]
While testing I found out that after this update graphviz can't create png files... Perhaps, some libraries should be recompiled. 

For example, for this file 

```
digraph CIG  {
  center=true; ratio=compress;
  rule0 [shape=record,label="<head> runtime:neighbor(?a,?b)|<00> livesIn(?a,?ra)|<01> livesIn(?b,?rb)|<02> regions:neighborRegions(?ra,?rb)"];
  query [shape=record,label="<head> |<00> rdfs:type(?x,Student)| <01> attendsLection (?x,?lx) | <02> isGivenBy(?lx, ?ll) | <03> runtime:neighbor(?ll,  ?x)"];

  query:03 -> rule0:head [label="S={ b_2=x_1%, x_1=x_1%, a_2=ll_1%, ll_1=ll_1% }; Q={}"];
}
```

output for "dot -Tpng cig.dot " is empty.
Do you see this in your environment?

**Comment by jimklimov** [comment]
Sorry, can not confirm or deny this issue: my graphviz build was not installed because as I see now "gmake publish" failed due to 

```
Couldn't find 'usr/share/graphviz/demo/modgraph.php' in any of the specified search directories:
...
Couldn't find 'usr/lib/graphviz/php/libgv_php.so' in any of the specified search directories:
...
Couldn't find 'usr/lib/graphviz/php/gv.php' in any of the specified search directories:
...
Couldn't find 'usr/share/man/man3/gv-php.3' in any of the specified search directories:
```

which I'd say is a bug in the graphviz recipe Makefile - if we require those build-products, we should require them explicitly and fail at `gmake build` step if we can not build them.

For now, I'm not sure what was missing for them to be built - installed PHP56 but it did not help.

I just copied over the build prototype's `usr` into my test zone root, so had it "fake-installed", and PNG generation seems to have worked; visually the generated images are identical:

```
jim@jimoi:~/shared/oi-userland/components/image/graphviz$ dot -Tgif test-pr1862.dot > test-pr1862.gif
jim@jimoi:~/shared/oi-userland/components/image/graphviz$ dot -Tpng test-pr1862.dot > test-pr1862.png
jim@jimoi:~/shared/oi-userland/components/image/graphviz$ ls -la test*
-rw-r--r-- 1 jim jim  441 Apr 20 09:15 test-pr1862.dot
-rw-r--r-- 1 jim jim 5377 Apr 20 09:27 test-pr1862.gif
-rw-r--r-- 1 jim jim 6031 Apr 20 09:27 test-pr1862.png
jim@jimoi:~/shared/oi-userland/components/image/graphviz$ file test*
test-pr1862.dot: HTML document, ASCII text
test-pr1862.gif: GIF image data, version 89a, 1309 x 181
test-pr1862.png: PNG image data, 1309 x 181, 8-bit/color RGB, non-interlaced

$ dot -V
dot - graphviz version 2.28.0 (20160414.0049)
```

Regarding the original issue, some googling suggested that graphviz can rely on other graphics libraries as well, so the relevant stack has to be all rebuilt and reinstalled (links from 2011-2013 note libgd and cairo in particular) when migrating to libpng16 from earlier major branches.

Building cairo went well for me, but `gmake publish` failed due to libOGL symlink being hijacked by VirtualBox, like with GLEW IIRC. This is something we have to tackle too, but separately I guess... I gather this is not an issue for Hipster build farm at least.

**Comment by pyhalov** [comment]
Retested. Seems recompiled graphviz (dot) generates correct png images. Need to test components compilation. These are:
pkg:/audio/sox
pkg:/consolidation/userland/userland-incorporation
pkg:/desktop/compiz
pkg:/developer/opensolaris/X
pkg:/editor/diagram/dia
pkg:/editor/gnu-emacs/gnu-emacs-gtk
pkg:/editor/gnu-emacs/gnu-emacs-x11
pkg:/entire
pkg:/games/openttd/openttd-core
pkg:/image/editor/gimp
pkg:/image/editor/inkscape
pkg:/image/imagemagick
pkg:/image/library/gegl
pkg:/image/library/libpng
pkg:/image/library/sdl-image
pkg:/image/rrdtool
pkg:/image/scanner/xsane
pkg:/image/viewer/gthumb
pkg:/image/xcursorgen
pkg:/library/audio/gstreamer/plugin/bad
pkg:/library/audio/gstreamer/plugin/good
pkg:/library/desktop/gdk-pixbuf
pkg:/library/desktop/search/tracker
pkg:/library/desktop/webkitgtk
pkg:/library/desktop/webkitgtk2
pkg:/library/e/efl
pkg:/library/gd
pkg:/library/graphics/wxwidgets
pkg:/library/libpoppler
pkg:/library/libwebp
pkg:/library/print/cups-libs
pkg:/library/qt4
pkg:/library/slang
pkg:/media/vlc
pkg:/metapackages/gui-install
pkg:/print/filter/ghostscript
pkg:/runtime/dosbox
pkg:/server_install
pkg:/system/library/freetype-2
pkg:/system/qemu/kvm
pkg:/web/analytics/webalizer
pkg:/web/php-54/extension/php-gd
pkg:/web/php-55/extension/php-gd
pkg:/web/php-56/extension/php-gd
pkg:/web/php-70/extension/php-gd

**Comment by pyhalov** [comment]
1) freetype should be built as first dependent component;
2) http://bugs.ghostscript.com/show_bug.cgi?id=692060 (ghostcript has to be patched)

**Comment by pyhalov** [comment]
Please, add REQUIRED_PACKAGES to the changed Makefiles and let's merge it...

**Comment by jimklimov** [comment]
Bumped the Makefiles

**Pull request merged**